internal/terminal/log.go

```diff
@@ -0,0 +1,115 @@
+package terminal
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/iancoleman/orderedmap"
+)
+
+// LogLevel is the level of a terminal log
+type LogLevel string
+
+// set of supported log levels
+const (
+	LogLevelInfo  LogLevel = "info"
+	LogLevelError LogLevel = "error"
+)
+
+var (
+	allLogLevels = []LogLevel{LogLevelInfo, LogLevelError}
+
+	longestLogLevel = func() LogLevel {
+		var longest LogLevel
+		for _, level := range allLogLevels {
+			if l := len(level); l > len(longest) {
+				longest = level
+			}
+		}
+		return longest
+	}()
+)
+
+// LogData produces the log data
+type LogData interface {
+	Message() (string, error)
+	Payload() ([]string, map[string]interface{}, error)
```

Comment:
couldn't this just be the struct and the error? looks like we are just calling `Marshal` on the object at the end of the day. We could simplify this to return an `interface{}, error` and call json.Marshal directly on that returned `interface{}`

Reply:
Also - do we even need to return an error? Looks like we can just return an `interface{}`

Reply:
there are two things this approach solves:

1. allow for each `LogData` type to specify its own fields to be marshaled (e.g. `TextMessage` produces a `message` field whereas `ErrorMessage` produces an `err` field
1. ensure a certain ordering of the fields in the JSON log so that its easy/convenient to the human eye

to account for the first goal, I'm marshaling a `map[string]interface{}` so that we can support a flat yet dynamic JSON field structure (no need for a nested `data` field would would _then_ have `message` or `err` present).  since maps aren't deterministic in order, I'm using the returned `[]string` to control that

I've worked around with this a lot and this seemed to be my only path to achieve the above, but let me know if you think I'm missing an alternative approach here

---

with regards to returning the error, so far only `nil` is returned so it could be removed, but the error in the function returns was an intentional include here so that it mirrored the function type of `Message()`. these both are "computed values", and there's not really anything special about `Payload` compared to `Message` that says its should be free-from-errors

if you feel strongly about it, i'll go ahead and remove.  but in the context of how these functions are consumed, it doesn't save any downstream complexity.